WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person Payments/CardPresentPaymentsOnboardingIPPUsersRefresher.swift

```diff
@@ -1,27 +1,41 @@
 import CoreData
 import Yosemite
 import Storage
+import Combine
 
 /// Refreshes the CPP onboarding state if there are IPP transactions stored
 ///
 class CardPresentPaymentsOnboardingIPPUsersRefresher {
     private let stores: StoresManager
     private let cardPresentPaymentsOnboardingUseCase: CardPresentPaymentsOnboardingUseCaseProtocol
+    private var cancellables: Set<AnyCancellable> = []
 
     init(stores: StoresManager = ServiceLocator.stores,
          cardPresentPaymentsOnboardingUseCase: CardPresentPaymentsOnboardingUseCaseProtocol = CardPresentPaymentsOnboardingUseCase()) {
         self.stores = stores
         self.cardPresentPaymentsOnboardingUseCase = cardPresentPaymentsOnboardingUseCase
     }
 
-    func refreshIPPUsersOnboardingState() {
+    func refreshIPPUsersOnboardingState(completion: @escaping (() -> Void)) {
         guard let siteID = stores.sessionManager.defaultStoreID else {
             return
         }
 
         let action = AppSettingsAction.loadSiteHasAtLeastOneIPPTransactionFinished(siteID: siteID) { [weak self] result in
             if result {
-                self?.cardPresentPaymentsOnboardingUseCase.refresh()
+                guard let self else { return }
+
+                self.cardPresentPaymentsOnboardingUseCase.statePublisher.sink { [weak self] state in
+                    guard state != .loading else {
+                        return
+                    }
+                    // Stop observing further state updates, to avoid memory cycles; this class is long-lived!
+                    _ = self?.cancellables.map { $0.cancel() }
```

Comment:
This might be problematic IMHO. If I understood correctly, we're cancelling all cancellables, not only the onboarding use case state updates. If we had another cancellable on this class that wasn't related to this action, it would be cancel, which might not be the intended effect.
Could we set a reference to the cancellables we want to cancel, and cancel only those?

Reply:
Yes, I agree – I thought about replacing the cancellables set with a single var, and only cancelling that, but that adds some risk that prematurely cancel a previous attempt; we don't want to do that, we want to adopt previous attempts where we can.

On further reflection of your suggestion, I think we should add a new `var onboardingStateObservations: Set<AnyCancellable>` and cancel all of those in this block. That should prompt anyone adding another cancellable to create a new set, but just in case it doesn't, I'll leave the existing `cancellables` in place and just not add anything to it (I believe double-cancelling something has undefined behaviour....)